lib/configurators/aws/aws.go

```diff
@@ -499,19 +504,64 @@ func policiesTarget(flags configurators.BootstrapFlags, accountID string, partit
 	if flags.AttachToRole != "" {
 		roleArn := flags.AttachToRole
 		if !arn.IsARN(flags.AttachToRole) {
-			roleArn = fmt.Sprintf("arn:%s:iam::%s:role/%s", partitionID, accountID, flags.AttachToRole)
+			roleArn = buildIAMARN(partitionID, accountID, "role", flags.AttachToRole)
 		}
 
 		return awslib.IdentityFromArn(roleArn)
 	}
 
 	if identity == nil {
-		return awslib.IdentityFromArn(fmt.Sprintf("arn:%s:iam::%s:user/%s", partitionID, accountID, defaultAttachUser))
+		return awslib.IdentityFromArn(buildIAMARN(partitionID, accountID, "user", defaultAttachUser))
+	}
+
+	if identity.GetType() == "assumed-role" {
+		roleIdentity, err := getRoleARNForAssumedRole(iamClient, identity)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		return roleIdentity, nil
 	}
 
 	return identity, nil
 }
 
+// buildIAMARN constructs an AWS IAM ARN string from the given partition,
+// account, resource type, and resource.
+// If the resource starts with the "/" prefix, this function takes care not to
+// add an additional "/" prefix to the constructed ARN.
+// This handles resource names that include a path correctly. Example:
+// resource input: "/some/path/to/rolename"
+// arn output: "arn:aws:iam::123456789012:role/some/path/to/rolename"
+func buildIAMARN(partitionID, accountID, resourceType, resource string) string {
+	if strings.HasPrefix(resource, "/") {
+		return fmt.Sprintf("arn:%s:iam::%s:%s%s", partitionID, accountID, resourceType, resource)
+	} else {
+		return fmt.Sprintf("arn:%s:iam::%s:%s/%s", partitionID, accountID, resourceType, resource)
+	}
+}
+
+// failedToResolveAssumeRoleARN is an error message returned when an
+// assumed-role identity cannot be resolved to the role ARN that it assumes,
+// which is necessary to attach policies to the identity.
+// Rather than returning errors about why it failed, this message suggests a
+// simple fix for the user to specify a role or user to attach policies to.
+const failedToResolveAssumeRoleARN = "Running with assumed-role credentials. Policies cannot be attached to an assumed-role. Provide the name or ARN of the IAM user or role to attach policies to."
+
+// getRoleARNForAssumedRole attempts to resolve assumed-role credentials to the underlying role ARN using IAM API.
+// This is necessary since the assumed-role ARN does not include the role path, so we cannot reliably reconstruct
+// the role ARN ourselves. When assumed-role credentials are used, we need the role ARN for
+func getRoleARNForAssumedRole(iamClient iamiface.IAMAPI, identity awslib.Identity) (awslib.Identity, error) {
+	roleOutput, err := iamClient.GetRole(&iam.GetRoleInput{RoleName: aws.String(identity.GetName())})
+	if err != nil || roleOutput == nil || roleOutput.Role == nil || roleOutput.Role.Arn == nil {
+		return nil, trace.BadParameter(failedToResolveAssumeRoleARN)
+	}
```

Comment:
I believe `GetRole` would require IAM permission to perform. As mentioned in the above comment, I don't think we need the ARN of the real role.